server/s3_copy_object_action.cc

```diff
@@ -82,6 +84,145 @@ void S3CopyObjectAction::setup_steps() {
+      std::bind(&S3CopyObjectAction::fetch_source_bucket_info_failed, this));
+
+  s3_log(S3_LOG_DEBUG, request_id, "Exiting\n");
+}
+
+void S3CopyObjectAction::fetch_source_bucket_info_success() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  s3_log(S3_LOG_DEBUG, request_id, "Found source bucket: [%s] metadata\n",
+         source_bucket_name.c_str());
+
+  // fetch source object metadata
+  fetch_source_object_info();
+
+  s3_log(S3_LOG_DEBUG, request_id, "Exiting\n");
+}
+
+void S3CopyObjectAction::fetch_source_bucket_info_failed() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+
+  s3_copy_action_state = S3CopyObjectActionState::validationFailed;
+
+  if (source_bucket_metadata->get_state() == S3BucketMetadataState::missing) {
+    s3_log(S3_LOG_DEBUG, request_id, "Source bucket: [%s] not found\n",
+           source_bucket_name.c_str());
+    set_s3_error("NoSuchBucket");
+  } else if (source_bucket_metadata->get_state() ==
+             S3BucketMetadataState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Source bucket metadata load operation failed due to pre launch "
+           "failure\n");
+    set_s3_error("ServiceUnavailable");
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id, "Source bucket metadata fetch failed\n");
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, request_id, "Exeting\n");
+}
+
+void S3CopyObjectAction::fetch_source_object_info() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  s3_log(S3_LOG_DEBUG, request_id, "Found source bucket metadata\n");
+  m0_uint128 source_object_list_oid =
+      source_bucket_metadata->get_object_list_index_oid();
+  m0_uint128 source_object_version_list_oid =
+      source_bucket_metadata->get_objects_version_list_index_oid();
+
+  if ((source_object_list_oid.u_hi == 0ULL &&
+       source_object_list_oid.u_lo == 0ULL) ||
+      (source_object_version_list_oid.u_hi == 0ULL &&
+       source_object_version_list_oid.u_lo == 0ULL)) {
+    // Object list index and version list index missing.
+    fetch_source_object_info_failed();
+  } else {
+    source_object_metadata =
+        object_metadata_factory->create_object_metadata_obj(
+            request, source_object_list_oid, source_bucket_name,
+            source_object_name);
+
+    source_object_metadata->set_objects_version_list_index_oid(
+        source_object_version_list_oid);
+
+    source_object_metadata->load(
+        std::bind(&S3CopyObjectAction::fetch_source_object_info_success, this),
+        std::bind(&S3CopyObjectAction::fetch_source_object_info_failed, this));
+  }
+  s3_log(S3_LOG_DEBUG, request_id, "Exiting\n");
+}
+
+void S3CopyObjectAction::fetch_source_object_info_success() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  s3_log(S3_LOG_DEBUG, request_id,
+         "Successfully fetched source object metadata\n");
+
+  if (MaxCopyObjectSourceSize < source_object_metadata->get_content_length()) {
```

Comment:
Do we really need this constraint?

Reply:
yes, CopyObject API only support source objects of size less than 5GB